src/context/Result.h

```diff
@@ -45,22 +49,44 @@ class Result final {
         return core_.iter->size();
     }
 
-    std::unique_ptr<Iterator> iter() const {
+    std::unique_ptr<Iterator> iter() const & {
         return core_.iter->copy();
     }
 
+    std::unique_ptr<Iterator> iter() && {
+        return std::move(core_.iter);
+    }
+
+    Iterator* iterRef() {
+        return core_.iter.get();
+    }
+
 private:
     friend class ResultBuilder;
     friend class ExecutionContext;
 
     Value&& moveValue() {
-        return std::move(*core_.value);
+        return std::move(*core_.values.front());
     }
 
     struct Core {
+        Core() = default;
+        Core(Core &&) = default;
+        Core& operator=(Core &&) = default;
+        Core(const Core &c) {
+            *this = c;
+        }
+        Core& operator=(const Core &c) {
+            state = c.state;
+            msg = c.msg;
+            values = c.values;
+            iter = c.iter->copy();
+            return *this;
+        }
+
         State state;
         std::string msg;
-        std::shared_ptr<Value> value;
+        std::vector<std::shared_ptr<Value>> values;
```

Comment:
Why we need a vector?

Reply:
Result should hold the all value it used, e.g. union sequential iterator will hold two independent value.

Reply:
I don't see any usage in current pr. Besides, the union does not hold  two value.

Reply:
Ok, the implementation of union had be modified after this pr. I will check this usage later.

Reply:
Fixed.